spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/ConfigMapPropertySource.java

```diff
@@ -89,12 +95,43 @@ private static String getName(KubernetesClient client, String name,
 	}
 
 	private static Map<String, Object> getData(KubernetesClient client, String name,
-			String namespace, Environment environment) {
+			String namespace, Environment environment, boolean enableVersioning) {
 		try {
 			Map<String, Object> result = new LinkedHashMap<>();
-			ConfigMap map = StringUtils.isEmpty(namespace)
-					? client.configMaps().withName(name).get()
-					: client.configMaps().inNamespace(namespace).withName(name).get();
+			ConfigMap map = null;
+			if (enableVersioning) {
+				String version = environment.getProperty("info.app.version");
+				Optional<ConfigMap> optMap = StringUtils.isEmpty(namespace)
+						? client.configMaps().list().getItems().stream()
+								.filter(configMap -> configMap.getMetadata().getLabels()
+										.containsKey(LABEL_APP)
+										&& configMap.getMetadata().getLabels()
+												.get(LABEL_APP).equals(name)
+										&& configMap.getMetadata().getLabels()
+												.containsKey(LABEL_VERSION)
+										&& configMap.getMetadata().getLabels()
+												.get(LABEL_VERSION).equals(version))
+								.findFirst()
+						: client.configMaps().inNamespace(namespace).list().getItems()
+								.stream()
+								.filter(configMap -> configMap.getMetadata().getLabels()
+										.containsKey(LABEL_APP)
+										&& configMap.getMetadata().getLabels()
+												.get(LABEL_APP).equals(name)
+										&& configMap.getMetadata().getLabels()
+												.containsKey(LABEL_VERSION)
+										&& configMap.getMetadata().getLabels()
+												.get(LABEL_VERSION).equals(version))
+								.findFirst();
+				if (optMap.isPresent()) {
```

Comment:
what happens if it is not present?

Reply:
it returns null, the same as for example here `client.configMaps().withName(name).get()` if it not found matching configmap